### ISSUE TITLE
[learning] Learn hierarchical candidate-family distributions and bounded residual pose proposals

### ISSUE THREAD

**Issue by JanDuchscherer104** [JanDuchscherer104/ARIA-NBV#76]
Parent: #67

Depends on: #68–#74 for deterministic state/context, typed families, reservoirs, feasibility, benchmark, and online oracle; #75 should establish a calibrated fixed-table selector first.

## Outcome

Implement a learned **candidate proposal policy** that adapts semantic family allocation and bounded pose parameters to actor-visible state, while retaining fixed handcrafted anchors, hard feasibility, exact behavior-policy provenance, and final validation by the existing target-RRI oracle.

The first learned proposer should not emit arbitrary world-frame SE(3) poses. It should control interpretable family distributions and optionally refine stable anchors with bounded 5-DoF residuals.

## Why this is a distinct problem

A selector learns

```math
\pi_\theta(a\mid s,Q),
```

over a supplied table. A proposer learns the support itself:

```math
Q \sim p_\phi(Q\mid s,e,H,b).
```

Changing `p_phi` changes which actions can be discovered and therefore changes the scientific action contract. Proposal quality must be evaluated separately through #73 before comparing downstream selectors.

## External orientation

GenNBV “extends typical limited action space to 5D free space,” demonstrating that learned continuous NBV is feasible, but its coverage reward and drone simulator are not ARIA-NBV's target-RRI setting.

Advantage-Weighted Regression offers a lower-risk first training route because it can “regress onto weighted target actions for the policy” using off-policy/dense oracle data rather than requiring immediate end-to-end on-policy RL.

## Proposed hierarchy

Factor the learned proposal distribution into semantic and continuous parts:

```math
k_i \sim \operatorname{Cat}(\alpha_\phi(s,e,H,b)),
```

```math
z_i \sim p_\phi(z\mid k_i,s,e,H,b),
```

```math
q_i = G_{k_i}(z_i; s,e) \oplus \Delta q_{\phi,i}.
```

Where:

- `k_i` selects an interpretable center/gaze/path family;
- `z_i` contains bounded family parameters such as radius, orbit angle, stand-off, lateral offset, yaw, and pitch;
- `G_k` is the audited semantic constructor from #69;
- `Delta q` is an optional bounded local 5-DoF residual;
- hard feasibility from #72 is applied after construction;
- proposal lineage and log probabilities are retained before and after final-shell selection.

## Actor inputs

The proposer may consume only actor-admissible state:

```text
observed/predicted target descriptor
actor scene/support carrier
current/reference pose
selected pose and directional history
step, horizon, and remaining budget
previous family survival/support summaries
candidate-generation contract and requested shell size
```

Forbidden inputs:

```text
GT mesh/crop
oracle RRI or target errors
all-candidate rendered depth/points
future observations
GT match identity as an actor feature
```

Oracle information may supervise the proposer offline or provide rewards through #74.

## Required policy API

```python
class CandidateProposalPolicy(Protocol):
    def propose(
        self,
        observation: ActorProposalObservation,
        request: CandidateProposalRequest,
        *,
        generator: torch.Generator,
    ) -> LearnedProposalBatch: ...

@dataclass
class LearnedProposalBatch:
    family_logits: Tensor
    family_probabilities: Tensor
    family_indices: Tensor
    continuous_parameters: Tensor
    conditional_log_probabilities: Tensor
    proposal_log_probabilities: Tensor
    anchor_mask: Tensor
    residuals_5d: Tensor | None
    proposed_poses_world: PoseTW
    checkpoint_hash: str
    model_config_hash: str
```

A downstream adapter applies semantic construction, hard feasibility, reservoir/final-shell selection, and persistence. Do not let the network bypass frame, camera, or validity owners.

## Fixed-anchor exploration contract

During training and initial evaluation, reserve a manifest-defined fraction of the final table for fixed audited anchors, initially around 20–40%:

```text
empirical Aria motion anchors
forward/body-continuity anchors
target orbit/stand-off anchors
turn-in-place anchors
one broad/free-shell exploration control
```

The learned proposer fills the remaining rows. Fixed anchors:

- prevent early proposal collapse;
- retain support outside the current policy's mode;
- provide a stable comparison/control;
- keep online data collection recoverable when the learned distribution becomes poor.

The exact anchor fraction is a knob and must be persisted.

### Stage 1 — oracle-elite / advantage-weighted imitation

Use dense reference-reservoir labels from #73. Let `r_i` be valid-row target gain and `b(s)` a state baseline. Define clipped weights:

```math
w_i = \operatorname{clip}
\left(\exp\frac{r_i-b(s)}{\tau},0,w_{max}\right).
```

Train family and continuous distributions by weighted log likelihood:

```math
\mathcal L_{prop}
= -\frac{\sum_i m_i^{train}w_i\log p_\phi(q_i\mid s,e)}
         {\sum_i m_i^{train}w_i+\epsilon}.
```

Options:

- top-quantile elite cloning as the simplest control;
- AWR with state/target baselines;
- family-balanced weights so common forward rows do not dominate;
- separate feasibility loss on attempted invalid rows.

Invalid actions never receive fabricated low RRI. They supervise feasibility/constraint prediction only.

### Stage 2 — set-level proposal objective

A proposal set should contain at least one strong action and remain diverse. Candidate-level likelihood alone may collapse onto one mode. Add controlled regularizers/diagnostics:

```text
family entropy
continuous entropy
pairwise SE(3)/target-direction repulsion
minimum fixed-family support
best-of-K surrogate or elite recall
```

Do not optimize diversity as a substitute for RRI; it is a support regularizer.

### Stage 3 — bounded residual proposal

Start from an anchor pose and predict local residuals:

```math
\Delta q = (\Delta x,\Delta y,\Delta z,\Delta\psi,\Delta\theta),
```

with `tanh`/distribution bounds tied to the family contract. Roll remains fixed in the canonical 5-DoF profile.

Residuals are expressed in an explicit root-, path-, or target-local frame. World-frame residual outputs are forbidden.

### Stage 4 — optional online updates

Collect through #74:

1. proposer emits anchors + learned rows and exact log probabilities;
2. hard feasibility/final-shell selection produces the action table;
3. fixed or learned selector chooses an action;
4. oracle scores dense/subset/selected evidence;
5. transitions and proposal records enter replay;
6. proposer updates in bounded collection/training phases.

Initial online algorithms may use:

- continued AWR from newly collected dense labels;
- REINFORCE/actor-critic for selected-only rewards;
- conservative off-policy updates using stored behavior densities;
- active oracle querying for uncertain/diverse rows.

Do not begin with PPO if dense oracle labels are already available and cheaper supervised objectives remain unexhausted.

## Behavior-policy provenance

Persist for every learned attempted row:

```text
proposal checkpoint content hash
model/config/training-manifest hash
family logits/probabilities and sampled family
conditional distribution parameters
proposal log probability
sampling noise/seed key
anchor versus learned flag
pre-feasibility pose
post-construction pose
residual frame and value
reservoir/final-shell inclusion lineage
online collection/update generation
```

If deterministic final-shell diversity selection prevents an exact final inclusion probability, retain the preselection proposal density and deterministic selector hash; do not invent a probability.

## Feasibility learning

Add a separate feasibility head or model output trained on every non-padding attempted row:

```math
\mathcal L_{feas} = \operatorname{BCE}(\hat p_{valid},m).
```

RRI/value loss remains valid-row-only. During early deployment/simulation, the hard geometric mask remains authoritative. A calibrated predicted-validity-times-value ranking is a later ablation, not a replacement for known hard constraints.

## Evaluation

Compare on frozen benchmark states:

1. fixed canonical 60-row profile;
2. fixed anchors plus untrained/random learned allocation control;
3. learned family allocator without residuals;
4. learned allocator plus residuals;
5. optional oracle-conditioned proposer upper reference.

Primary proposal metrics:

- proposal regret/best-of-60;
- reference top-set recall;
- valid and label-supported fraction;
- target-aware family survival;
- family and directional entropy;
- near-duplicate rate;
- endpoint gain under the same selector;
- path/feasibility failure;
- oracle queries, runtime, and memory.

A learned proposer is promoted only if gains survive hard-oracle evaluation and do not arise from oracle-input leakage or relaxed feasibility.

## Acceptance criteria

- [ ] A learned hierarchical family/parameter policy is implemented behind a typed proposal interface.
- [ ] It uses actor-visible state only and emits root/target-local bounded parameters.
- [ ] Fixed audited anchors occupy a persisted nonzero fraction during initial training/evaluation.
- [ ] Exact behavior proposal log probabilities are stored when defined.
- [ ] Oracle-elite and AWR controls train from dense reservoir labels.
- [ ] Invalid rows train feasibility only; RRI/value losses remain valid-row-only.
- [ ] Optional residuals are bounded 5-DoF local corrections.
- [ ] Reservoir/final-shell selection preserves learned-row provenance.
- [ ] Candidate-generator evaluation uses #73 and hard-oracle validation.
- [ ] Downstream selector comparisons freeze either the proposer checkpoint or the realized candidate table.
- [ ] Online updates are checkpoint/version bounded and reproducible.

## Required tests

- actor/oracle dependency and target-source dropout;
- family categorical/log-probability correctness;
- continuous distribution bounds and local-frame transforms;
- anchor fraction and row identity;
- residual zero-equivalence to anchors;
- behavior density persistence/round trip;
- learned attempted rows through hard feasibility and final-shell mapping;
- invalid-row loss isolation;
- deterministic sampling under keyed generators;
- proposal collapse fixtures and entropy/family-floor behavior;
- stale checkpoint/contract rejection;
- hard-oracle benchmark/evaluation integration.

## External sources

- Chen et al., *GenNBV: Generalizable Next-Best-View Policy for Active 3D Reconstruction*: https://openaccess.thecvf.com/content/CVPR2024/html/Chen_GenNBV_Generalizable_Next-Best-View_Policy_for_Active_3D_Reconstruction_CVPR_2024_paper.html
- Peng et al., *Advantage-Weighted Regression*: https://arxiv.org/abs/1910.00177

## Non-goals

- Do not generate unconstrained world-frame SE(3) actions in the first version.
- Do not remove fixed anchors until support/collapse evidence justifies it.
- Do not optimize against a learned surrogate without hard-oracle re-evaluation.
- Do not present online continuous-policy results as part of the thesis core unless the scope and evidence protocol are explicitly revised.

**Comment by JanDuchscherer104** [comment]
## Research enrichment — round 2: add black-box CEM controls, exact parameter-space densities, and freeze alternating learners

### 1. Add Cross-Entropy Method as the first continuous-proposal control

Before PPO or differentiable pose refinement, implement a bounded black-box optimizer over the semantic family parameters. The Cross-Entropy Method iteratively fits a proposal distribution to elite samples and is designed for difficult continuous/nonconvex optimization:

- Rubinstein, *The Cross-Entropy Method for Combinatorial and Continuous Optimization*: https://doi.org/10.1023/A:1010091220143
- de Boer et al., tutorial: https://doi.org/10.1007/s10479-005-5724-z

Use CEM in two roles:

```text
CEM_SURROGATE
  many cheap queries to a frozen learned scorer, final hard-oracle validation

CEM_ORACLE_BOUNDED
  very small query-budget upper-reference on benchmark/pilot states only
```

CEM provides a strong non-gradient continuous baseline and can generate elite proposal targets for AWR/distillation. It must report query count and cannot silently consume unlimited oracle calls.

### 2. Define the stochastic action in parameter space

For semantic family `k` and bounded local parameters `z`, the policy samples `(k,z)`, then the deterministic constructor maps to pose `q=G_k(z;s,e)`.

Persist and optimize:

```math
\log p_\phi(k,z\mid s,e),
```

not an ill-defined world-pose density unless the constructor is bijective and its Jacobian is handled. The stored behavior probability must state:

```text
family categorical semantics
continuous parameter distribution
bounds/transforms and clipping
sampling with/without replacement
joint versus per-row log probability
constructor version
pose-density/Jacobian = available | unavailable
```

For a set of `N` proposals, the joint probability is not generally the sum of independent row terms when family counts, without-replacement draws, quotas, or diversity selection couple rows.

### 3. Separate attempted proposal likelihood from final-shell selection

#71 may apply hard feasibility, quotas, and deterministic diversity selection. Keep gradients/log probabilities attached to the **attempted parameter samples**. Record final inclusion separately.

Do not apply REINFORCE to a final-shell action as though its probability were the preselection proposal density unless the complete stochastic selector probability is known.

### 4. Alternate frozen selector and proposer phases

Jointly updating selector and proposer creates a moving support and moving objective. Start with bounded alternating phases:

1. freeze canonical proposal profile; train/calibrate selector;
2. freeze selector; train proposer from hard-oracle/reference-reservoir labels;
3. freeze proposer checkpoint; regenerate a versioned corpus;
4. retrain/evaluate selector on matched realized tables;
5. optionally repeat as a new acquisition round.

Every phase receives a new contract/checkpoint ID. Confirmatory evaluation freezes both proposer and selector.

### 5. Preserve broad support with anchors and entropy floors

Fixed anchors should be selected by role, not just percentage:

```text
empirical local motion
left/right orbit
stand-off correction
turn-in-place
a broad support control
```

Add minimum family probability/final counts and monitor effective sample size and pairwise diversity. Entropy alone can be high while all samples occupy one geometric mode.

### 6. Add active oracle-query design

Dense oracle scoring of every learned reservoir may be too costly. Compare:

```text
random anchor queries
surrogate top-k
surrogate uncertainty
geometric diversity/core-set queries
mixed top-k + uncertainty + random anchors
```

The query policy is separate from the proposal policy and must not see unseen oracle labels. Use #74/#81 provenance and report oracle labels per improvement.

### 7. Add proposal-distribution shift and collapse diagnostics

Monitor by state stratum:

```text
family entropy and minimum mass
parameter entropy / covariance eigenvalues
SE(3) duplicate rate
validity and final-inclusion rate
oracle top-set recall
proposal regret
anchor-versus-learned selection share
KL to previous checkpoint and to handcrafted control
```

A proposer that raises mean gain by collapsing on easy/large targets but worsens rare target strata should not be promoted.

### 8. Add off-policy safeguards only with valid behavior semantics

For online proposal updates, importance ratios require compatible attempted-action densities. Use clipping and report effective sample size. If deterministic final-shell selection or old schemas prevent exact behavior density, prefer AWR/elite imitation with explicit dataset provenance over pretending to have exact policy-gradient ratios.

### 9. Test provenance shortcuts in downstream selectors

A learned family allocation can make family ID an even stronger label shortcut. Evaluate the downstream selector with:

```text
family ID dropped
family ID randomized consistently with physical geometry controls
held-out proposer checkpoint/profile
same pose generated by anchor versus learned row
```

### Additional acceptance criteria

- [ ] CEM surrogate and bounded-oracle controls are implemented before claiming a need for end-to-end RL or pose gradients.
- [ ] Behavior log probabilities are defined in semantic parameter space with exact transform semantics.
- [ ] Coupled set sampling and deterministic final-shell inclusion are not mislabeled as independent pose probabilities.
- [ ] Proposer and selector are first trained in alternating frozen phases with versioned corpora.
- [ ] Fixed anchors guarantee named geometric support roles.
- [ ] Oracle-query policy is separate, budgeted, and provenance-complete.
- [ ] Collapse and rare-stratum diagnostics gate promotion.
- [ ] Off-policy correction is used only when behavior densities are genuinely recoverable.
- [ ] Held-out proposer/profile and provenance-shortcut tests are required.
